editor/osm_feature_matcher.cpp

```diff
@@ -49,54 +68,78 @@ vector<m2::PointD> GetWaysGeometry(pugi::xml_document const & osmResponse,
   return result;
 }
 
+vector<m2::PointD> GetRelationsGeometry(pugi::xml_document const & osmResponse,
+                                        pugi::xml_node const & relation)
+{
+  vector<m2::PointD> result;
+  ForEachRelationsNode(osmResponse, relation, [&result](XMLFeature const & xmlFt)
+                       {
+                         result.push_back(xmlFt.GetMercatorCenter());
+                       });
+  return result;
+}
+
+// TODO(mgsergio): XMLFeature should have GetGeometry method.
+vector<m2::PointD> GetWaysOrRelationsGeometry(pugi::xml_document const & osmResponse,
+                                              pugi::xml_node const & wayOrRelation)
+{
+  if (strcmp(wayOrRelation.name(), "way") == 0)
+    return GetWaysGeometry(osmResponse, wayOrRelation);
+  return GetRelationsGeometry(osmResponse, wayOrRelation);
+}
+
 /// @returns value form [-0.5, 0.5]. Negative values are used as penalty,
 /// positive as score.
-double ScoreGeometry(pugi::xml_document const & osmResponse, pugi::xml_node const & way,
-                     vector<m2::PointD> geometry)
+/// @param osmResponse - nodes, ways and relations from osm
+/// @param wayOrRelation - either way or relation to be compared agains ourGeometry
+/// @param outGeometry - geometry of a FeatureType
+double ScoreGeometry(pugi::xml_document const & osmResponse,
+                     pugi::xml_node const & wayOrRelation, vector<m2::PointD> ourGeometry)
 {
+  ASSERT(!ourGeometry.empty(), ("Our geometry cannot be empty"));
   int matched = 0;
 
-  auto wayGeometry = GetWaysGeometry(osmResponse, way);
+  auto theirGeometry = GetWaysOrRelationsGeometry(osmResponse, wayOrRelation);
+
+  if (theirGeometry.empty())
+    return -1;
 
-  sort(begin(wayGeometry), end(wayGeometry));
-  sort(begin(geometry), end(geometry));
+  my::SortUnique(theirGeometry);
+  my::SortUnique(ourGeometry);
```

Comment:
Почему он может поменяться? Если там не будет фильтрации дубликатов, то весь код ниже не будет работать правильно.

> On Mar 21, 2016, at 15:19, mgsergio <notifications@github.com> wrote:
> 
> In editor/osm_feature_matcher.cpp <https://github.com/mapsme/omim/pull/2372#discussion_r56813011>:
> 
> >  
> > -  sort(begin(wayGeometry), end(wayGeometry));
> > -  sort(begin(geometry), end(geometry));
> > +  my::SortUnique(theirGeometry);
> > +  my::SortUnique(ourGeometry);
> Это правда, но так надежней.
> Тот конец может поменяться.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub <https://github.com/mapsme/omim/pull/2372/files/904ccf66b255cd28dd6543a812461901ba00c06c#r56813011>